### ISSUE TITLE
Assignment 1, kNN classifier

### ISSUE THREAD

**Issue by truehines** [jkarimi91/cs231#1]
jkarimi91,

I think I have found an error in the k-fold cross validation snippet of the knn jupyter notebook. In the second segment of this snippet, the line:
`X_train_set = np.vstack(X_train_folds[:i] + X_train_folds[i + 1:])`

I believe that using the "+" operator on these two arrays (X_train_folds[:i] & X_train_folds[i+1:]) will actually add together the array elements instead of concatenating them as you intended. Do you agree with this?

In my own implementation I have the following (the reason for the if-elif-else is that concatenating an empty array gives an error):
`if i == 0:`
`X_train_fold = np.concatenate(X_train_folds[(i + 1):num_folds])`
`y_train_fold = np.concatenate(y_train_folds[(i + 1):num_folds])`
`elif i == (num_folds - 1):`
`X_train_fold = np.concatenate(X_train_folds[0:i])`
`y_train_fold = np.concatenate(y_train_folds[0:i])`
`else:`
`X_train_fold = np.concatenate((np.concatenate(X_train_folds[0:i]), np.concatenate(X_train_folds[(i + 1):num_folds])))`
`y_train_fold = np.concatenate((np.concatenate(y_train_folds[0:i]), np.concatenate(y_train_folds[(i + 1):num_folds])))`
`classifier.train(X_train_fold, y_train_fold)`

I am open to suggestions on a cleaner way to implement this...

Your feedback is greatly appreciated -- I don't have someone to discuss this type of thing with....

Regards,
True

**Comment by jkarimi91** [comment]
It is true that using the `+` operator on two arrays will do an element-wise addition. However, `X_train_folds` is a **list**. As a result, using the `+` operator on two lists will **concatenate** them. 
That concatenated list is then passed as an argument to `np.vstack`. Since the concatenated list is a list of arrays, `np.vstack` will create a new array by vertically stacking each of the arrays in the list:
```
concat_list = X_train_folds[:i] + X_train_folds[i + 1:]  # [array, array, array, ...]
X_train_fold = np.vstack(concat_list)
```

In other words, `X_train_set = np.vstack(X_train_folds[:i] + X_train_folds[i + 1:])` is correct.
